libs/agno/agno/agent/agent.py

```diff
@@ -2134,169 +2158,196 @@ def _convert_response_to_structured_format(self, run_response: RunResponse):
+                yield self.create_run_response(
+                    content=call_result.content,
+                    event=RunEvent.tool_call_completed,
+                    session_id=session_id,
+                )
+        if len(function_call_results) > 0:
+            run_messages.messages.extend(function_call_results)
+
     def _handle_tool_call_updates(self, run_response: RunResponse, run_messages: RunMessages):
         self.model = cast(Model, self.model)
         for _t in run_response.tools or []:
             # Case 1: Handle confirmed tools and execute them
             if _t.requires_confirmation is not None and _t.requires_confirmation is True and self._functions_for_model:
-                for func_name, func in self._functions_for_model.items():
-                    if func_name == _t.tool_name:
-                        func.requires_confirmation = False
                 # Tool is confirmed and hasn't been run before
                 if _t.confirmed is not None and _t.confirmed is True and _t.result is None:
-                    # Execute the tool
-                    function_call = self.model.get_function_call_to_run_from_tool_execution(
-                        _t, self._functions_for_model
-                    )
-                    function_call_results: List[Message] = []
-
-                    for call_result in self.model.run_function_call(
-                        function_call=function_call,
-                        function_call_results=function_call_results,
-                    ):
-                        if (
-                            call_result.event == ModelResponseEvent.tool_call_completed.value
-                            and call_result.tool_executions
-                        ):
-                            tool_execution = call_result.tool_executions[0]
-                            _t.result = tool_execution.result
-                            _t.tool_call_error = tool_execution.tool_call_error
-                    if len(function_call_results) > 0:
-                        run_messages.messages.extend(function_call_results)
+                    deque(self._run_tool(run_messages, _t), maxlen=0)
                     _t.requires_confirmation = False
                 else:
                     raise ValueError(f"Tool {_t.tool_name} requires confirmation, cannot continue run")
 
+            # Case 2: Handle external execution required tools
+            if _t.external_execution_required is not None and _t.external_execution_required is True:
+                self._handle_external_execution_update(run_messages=run_messages, tool=_t)
+
+            # Case 3: Handle user input required tools
+            if _t.requires_user_input is not None and _t.requires_user_input is True:
+                self._handle_user_input_update(tool=_t)
+                _t.requires_user_input = False
+                deque(self._run_tool(run_messages, _t), maxlen=0)
```

Comment:
Comment here? I had to check what was up

Reply:
Done